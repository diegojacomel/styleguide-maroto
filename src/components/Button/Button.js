// Modules
import React from "react";
import PropTypes from 'prop-types';

// Components
import Icon from '../Icon/Icon';

// Styles
import styled from "styled-components";

const OutlineButton = (theme, outline, color) => {
    if (outline) {
        return `
            border: 1px solid ${color};
            background-color: ${theme.colors.white};
            color: ${color};
            &:hover {
                border: 1px solid ${theme.colors.white};
                background-color: ${color};
                color: ${theme.colors.white};
            }
    `;
    } else {
        return `
            background-color: ${color};
            color: ${theme.colors.white};
    `;
    }
};

const SizeButton = (props) => {
    switch (props.size) {
        case 'xg':
            return `
                padding: ${props.theme.spacing.lg} ${props.theme.spacing.xg};
            `
        case 'lg':
            return `
                padding: ${props.theme.spacing.md} ${props.theme.spacing.xlg};
            `
        case 'md':
            return `
                padding: ${props.theme.spacing.sm} ${props.theme.spacing.lg};
            `
        case 'sm':
            return `
                padding: ${props.theme.spacing.xsm} ${props.theme.spacing.md};
            `
        case 'xs':
            return `
                padding: ${props.theme.spacing.xs} ${props.theme.spacing.sm};
            `
        default:
            return `
                padding: ${props.theme.spacing.sm} ${props.theme.spacing.lg};'
        `
    }
}

const ButtonStyle = styled.button`
    text-align: center;
    cursor: pointer;
    outline: none;
    border: none;
    line-height: 1;
    transition: all 0.3s;
    margin-right: 15px;
    white-space: nowrap;
    width: ${({ block }) => block ? '100%' : 'initial'};
    border-radius: ${props => props.theme.rounded[props.rounded]};
    font-size: ${props => props.theme.fontSize[props.size]};
    ${props => OutlineButton(props.theme, props.outline, props.theme.colors[props.color])};
    ${props => SizeButton(props)};
    &:hover {
        ${props => OutlineButton(props.theme, props.outline, props.theme.colors.hover[props.color])};
    }
    &:last-child {
        margin-right: 0;
    }
    & > span {
        display: flex;
        justify-content: space-between;
        align-items: center;
        flex-direction: ${props => props.iconLeft ? 'row-reverse' : 'row'};
        & > svg {
            margin-left: ${props => !props.iconLeft ? '10px' : '0'};
            margin-right: ${props => props.iconLeft ? '10px' : '0'};
        }
    }
`;

const Button = ({ children, type, color, outline, size, rounded, block, icon, iconSize, iconColor, iconLeft, disabled, onClick }) => (
    <ButtonStyle
        type={type}
        color={color}
        outline={outline}
        size={size}
        rounded={rounded}
        block={block}
        iconLeft={iconLeft}
        onClick={onClick}
        disabled={disabled}
        data-testid="button"
    >
        <span>
            {children}
            {icon
                ?
                <Icon tag={icon} iconSize={iconSize} iconColor={iconColor} />
                :
                null
            }
        </span>
    </ButtonStyle>
);

Button.propTypes = {
    type: PropTypes.string,
    color: PropTypes.string,
    outline: PropTypes.bool,
    size: PropTypes.string,
    rounded: PropTypes.string,
    block: PropTypes.bool,
    icon: PropTypes.string,
    iconSize: PropTypes.string,
    iconColor: PropTypes.string,
    disabled: PropTypes.bool,
    onClick: PropTypes.func
}

Button.defaultProps = {
    type: 'button',
    color: 'default',
    outline: false,
    size: 'md',
    rounded: 'none',
    block: false,
    disabled: false
}

export default Button;