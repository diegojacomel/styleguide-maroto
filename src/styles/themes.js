// Scales
import { pxToRem } from '../utils/scales';

const theme = {
    colors: {
        primary: "#54a0ff",
        danger: "#ff6b6b",
        info: "#48dbfb",
        success: "#1dd1a1",
        default: "#576574",
        white: "#ffffff",
        black: "#000000",
        greyLight: "#f2f2f2",
        greyMedium: "#999999",
        greyDark: "#333333",
        hover: {
            primary: "#1a81ff",
            danger: "#ff3333",
            info: "#05cdfa",
            success: "#159d79",
            default: "#373f49",
            white: "#f2f2f2",
            black: "#1a1a1a"
        }
    },
    spacing: {
        none: "0",
        xxs: pxToRem(3),
        xs: pxToRem(5),
        xsm: pxToRem(8),
        sm: pxToRem(10),
        md: pxToRem(15),
        lg: pxToRem(20),
        xlg: pxToRem(25),
        xg: pxToRem(30),
        xxg: pxToRem(50)
    },
    fontSize: {
        none: "0",
        xs: pxToRem(12),
        sm: pxToRem(14),
        md: pxToRem(16),
        lg: pxToRem(18),
        xg: pxToRem(20),
    },
    rounded: {
        none: "0",
        sm: "3px",
        md: "5px",
        lg: "10px",
        full: "100%"
    }
};

export {
    theme
};