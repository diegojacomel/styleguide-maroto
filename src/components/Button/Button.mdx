---
name: Button
route: /button
---

import Button from './Button'
import { Playground } from 'docz'
import { ThemeProvider } from 'styled-components'
import { theme } from '../../styles/themes';

# Button

## Basic usage

<ThemeProvider theme={theme}>
    <Button color="default" onClick={() => { console.log(theme) }}>Click here</Button>
</ThemeProvider>

```js
<Button>Click here</Button>
```

## Props

#### **type**

The type attribute specifies the type of button.

type: string

#### **color**

It define the button color.

type: string

- primary
- danger
- info
- success
- default

<Playground>
    <ThemeProvider theme={theme}>
        <div>
            <Button color="primary">primary</Button>
            <Button color="danger">danger</Button>
            <Button color="info">info</Button>
            <Button color="success">success</Button>
            <Button color="default">default</Button>
        </div>
    </ThemeProvider>
</Playground>

#### **outline**

It define if the button is outline or filled.

type: bool

<Playground>
    <ThemeProvider theme={theme}>
        <div>
            <Button color="primary" outline>primary</Button>
            <Button color="danger" outline>danger</Button>
            <Button color="info" outline>info</Button>
            <Button color="success" outline>success</Button>
            <Button color="default" outline>default</Button>
        </div>
    </ThemeProvider>
</Playground>

#### **size**

It define the button size.

type: string

- xg
- lg
- md
- sm
- xs

<Playground>
    <ThemeProvider theme={theme}>
        <div>
            <Button color="primary" size="xg">Button XG</Button>
            <Button color="primary" size="lg">Button LG</Button>
            <Button color="primary" size="md">Button MD</Button>
            <Button color="primary" size="sm">Button SM</Button>
            <Button color="primary" size="xs">Button XS</Button>
        </div>
    </ThemeProvider>
</Playground>

#### **rounded**

It define the button rounded size.

type: string

- none
- sm
- md
- lg
- full

<Playground>
    <ThemeProvider theme={theme}>
        <Button color="primary" rounded="lg">Button Rounded</Button>
    </ThemeProvider>
</Playground>

#### **block**

Let the button with full width display.

type: bool

<Playground>
    <ThemeProvider theme={theme}>
        <Button color="primary" block>Button Block</Button>
    </ThemeProvider>
</Playground>

#### **icon**

If you to wish putting a icon on button.

type: string

P.S.: To more information, you should find the Icon component on styleguide sidebar.

<Playground>
    <ThemeProvider theme={theme}>
        <Button color="primary" icon="arrow-right2" iconSize="md" iconColor="white">Button Icon</Button>
    </ThemeProvider>
</Playground>

#### **iconSize**

It define the icon size.

type: string

- xxs
- xs
- xsm
- sm
- md
- lg
- xlg
- xg
- xxg

#### **iconColor**

It define the button color.

type: string

- primary
- danger
- info
- success
- default

#### **iconLeft**

It should be used when you want reverse the icon side.

type: bool

#### **onClick**

It should be used to fire any method that you want.

type: function