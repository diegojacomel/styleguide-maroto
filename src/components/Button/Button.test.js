// Modules
import React from 'react';
import { render, fireEvent, cleanup } from 'react-testing-library';
import { ThemeProvider } from 'styled-components';
import { theme } from '../../styles/themes';

// Components
import Button from './Button';

// Unmount and cleanup DOM after the test is finished
afterEach(cleanup)

describe("Test Button component", () => {
    // Tests
    test("Clicked", () => {
        const { getByTestId, } = render(
            <ThemeProvider theme={theme}>
                <Button>Click here</Button>
            </ThemeProvider>
        );
        
        const element = getByTestId('button');

        fireEvent.click(element);
    });

    test("It's present in the document", () => {
        const { getByTestId, } = render(
            <ThemeProvider theme={theme}>
                <Button>Click here</Button>
            </ThemeProvider>
        );
        
        const element = getByTestId('button');

        expect(element).toBeInTheDocument();
    })

    test("It have attributes", () => {
        const { getByTestId, } = render(
            <ThemeProvider theme={theme}>
                <Button type="button" color="primary" size="md" rounded="md" onClick={() => {}} block iconLeft outline disabled>Click here</Button>
            </ThemeProvider>
        );
        
        const element = getByTestId('button');

        expect(element).toHaveAttribute('type');
        expect(element).toHaveAttribute('color');
        expect(element).toHaveAttribute('disabled');
    })

    test("Generate snapshots", () => {
        const { container, } = render(
            <ThemeProvider theme={theme}>
                <Button type="button" color="primary" size="md" rounded="md" onClick={() => {}} block>Click here</Button>
            </ThemeProvider>
        );

        expect(container.firstChild).toMatchSnapshot();
    })
})