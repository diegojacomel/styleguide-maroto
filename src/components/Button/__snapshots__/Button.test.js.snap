// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Button component Generate snapshots 1`] = `
<button
  class="sc-bwzfXH ccPxto"
  color="primary"
  data-testid="button"
  type="button"
>
  <span>
    Click here
  </span>
</button>
`;
